src/plugins/auto_batch/auto_batch.cpp

```diff
@@ -0,0 +1,718 @@
+        if (req->GetBlob(name) != blob)
+            req->SetBlob(name, blob);
+    }
+}
+
+void AutoBatchInferRequest::CopyInputsIfNeeded() {
+    for (const auto& it : _networkInputs) {
+        auto& name = it.first;
+        // this request is already in BUSY state, so using the internal functions safely
+        CopyBlobIfNeeded(GetBlob(name), _myBatchedRequestWrapper._inferRequestBatched->GetBlob(name), true);
+    }
+}
+
+void AutoBatchInferRequest::CopyBlobIfNeeded(InferenceEngine::Blob::CPtr src,
+                                             InferenceEngine::Blob::Ptr dst,
+                                             bool bInput) {
+    auto bufferDst = dst->buffer();
+    auto ptrDst = bufferDst.as<char*>();
+    auto bufferSrc = src->cbuffer();
+    auto ptrSrc = bufferSrc.as<const char*>();
+    ptrdiff_t szDst = dst->byteSize();
+    ptrdiff_t szSrc = src->byteSize();
+    if (bInput) {
+        ptrdiff_t offset = szSrc != szDst ? _batchId * szDst / _batchSize : 0;
+        if ((ptrDst + offset) == ptrSrc)
+            return;
+        else
+            memcpy(ptrDst + offset, ptrSrc, szSrc);
+    } else {
+        ptrdiff_t offset = szSrc != szDst ? _batchId * szSrc / _batchSize : 0;
+        if ((ptrSrc + offset) == ptrDst)
+            return;
+        else
+            memcpy(ptrDst, ptrSrc + offset, szDst);
+    }
+}
+
+void AutoBatchInferRequest::CopyOutputsIfNeeded() {
+    for (const auto& it : _networkOutputs) {
+        auto& name = it.first;
+        // this request is already in BUSY state, so using the internal functions safely
+        CopyBlobIfNeeded(_myBatchedRequestWrapper._inferRequestBatched->GetBlob(name), GetBlob(name), false);
+    }
+}
+
+std::map<std::string, InferenceEngine::InferenceEngineProfileInfo> AutoBatchInferRequest::GetPerformanceCounts() const {
+    return _perfMap;
+}
+
+AutoBatchAsyncInferRequest::AutoBatchAsyncInferRequest(
+    const AutoBatchInferRequest::Ptr& inferRequest,
+    const bool needPerfCounters,
+    InferenceEngine::SoIInferRequestInternal& inferRequestWithoutBatch,
+    const ITaskExecutor::Ptr& callbackExecutor)
+    : AsyncInferRequestThreadSafeDefault(inferRequest, nullptr, callbackExecutor),
+      _inferRequestWithoutBatch(inferRequestWithoutBatch),
+      _inferRequest{inferRequest} {
+    // this executor starts the inference while  the task (checking the result) is passed to the next stage
+    struct ThisRequestExecutor : public ITaskExecutor {
+        explicit ThisRequestExecutor(AutoBatchAsyncInferRequest* _this_) : _this{_this_} {}
+        void run(Task task) override {
+            auto& workerInferRequest = _this->_inferRequest->_myBatchedRequestWrapper;
+            std::pair<AutoBatchAsyncInferRequest*, InferenceEngine::Task> t;
+            t.first = _this;
+            t.second = std::move(task);
+            workerInferRequest._tasks.push(t);
+            // it is ok to call unsafe_size() here as the queue only grows (and the bulk removal happens under the
```

Comment:
Unsafe_size() should be called only in a safe context. For example when it is guaranteed that is safe to call this function.